substrate/runtime/contract/src/vm.rs

```diff
@@ -0,0 +1,769 @@
+//! to each other and operate on a simple key-value storage.
+
+use codec::Decode;
+use parity_wasm::elements::{self, External, MemoryType};
+use pwasm_utils;
+use pwasm_utils::rules;
+use rstd::prelude::*;
+use sandbox;
+use gas::{GasMeter, GasMeterResult};
+use runtime_primitives::traits::{As, CheckedMul};
+use {Trait};
+use exec::{CallReceipt, CreateReceipt};
+
+/// An interface that provides an access to the external environment in which the
+/// smart-contract is executed.
+///
+/// This interface is specialised to an account of the executing code, so all
+/// operations are implicitly performed on that account.
+pub trait Ext<T: Trait> {
+	/// Returns the storage entry of the executing account by the given key.
+	fn get_storage(&self, key: &[u8]) -> Option<Vec<u8>>;
+
+	/// Sets the storage entry by the given key to the specified value.
+	fn set_storage(&mut self, key: &[u8], value: Option<Vec<u8>>);
+
+	// TODO: Return the address of the created contract.
+	/// Create a new account for a contract.
+	///
+	/// The newly created account will be associated with the `code`. `value` specifies the amount of value
+	/// transfered from this to the newly created account.
+	fn create(
+		&mut self,
+		code: &[u8],
+		value: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		data: Vec<u8>,
+	) -> Result<CreateReceipt<T>, ()>;
+
+	/// Call (possibly transfering some amount of funds) into the specified account.
+	fn call(
+		&mut self,
+		to: &T::AccountId,
+		value: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		data: Vec<u8>,
+	) -> Result<CallReceipt, ()>;
+}
+
+/// Error that can occur while preparing or executing wasm smart-contract.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+	/// Error happened while serializing the module.
+	Serialization,
+
+	/// Error happened while deserializing the module.
+	Deserialization,
+
+	/// Internal memory declaration has been found in the module.
+	InternalMemoryDeclared,
+
+	/// Gas instrumentation failed.
+	///
+	/// This most likely indicates the module isn't valid.
+	GasInstrumentation,
+
+	/// Stack instrumentation failed.
+	///
+	/// This  most likely indicates the module isn't valid.
+	StackHeightInstrumentation,
+
+	/// Error happened during invocation of the contract's entrypoint.
+	///
+	/// Most likely because of trap.
+	Invoke,
+
+	/// Error happened during instantiation.
+	///
+	/// This might indicate that `start` function trapped, or module isn't
+	/// instantiable and/or unlinkable.
+	Instantiate,
+
+	/// Memory creation error.
+	///
+	/// This might happen when the memory import has invalid descriptor or
+	/// requested too much resources.
+	Memory,
+}
+
+/// Enumerates all possible *special* trap conditions.
+///
+/// In this runtime traps used not only for signaling about errors but also
+/// to just terminate quickly in some cases.
+enum SpecialTrap {
+	/// Signals that trap was generated in response to call `ext_return` host function.
+	Return(Vec<u8>),
```

Comment:
Just a note: we might want to avoid the extra copying here in the future. One way I can think of is to change this enum point to a region of the memory, and then like we do in `parity-ethereum`, directly wrap that memory as the return data.

Reply:
Yeah, I will add `TODO` about this!